### ISSUE TITLE
Protocol method only considers the first overload of the matched method, ignoring all others

### ISSUE THREAD

**Issue by jorenham** [microsoft/pyright#9663]
Here I expected that the `CanMul` protocol would match all of the `__mul__` overloads, so that when used in the `mul` function, `mul` would also behave as if it was overloaded that way:

```py
from typing import Callable, Never, Protocol, overload, reveal_type

class Two:
    @overload
    def __mul__(self, x: str, /) -> str: ...
    @overload
    def __mul__(self, x: int, /) -> int: ...
    def __mul__(self, x: str | int, /) -> str | int:
        return 2 * x

reveal_type(Two() * "a")  # Type of "Two() * "a"" is "str"
reveal_type(Two() * 1)  # Type of "Two() * 1" is "int"

###

class CanMul[X, Y](Protocol):
    def __mul__(self, x: X, /) -> Y: ...

def mul[X, Y](lhs: CanMul[X, Y], rhs: X, /) -> Y:
    return lhs * rhs

reveal_type(mul(Two(), "a"))  # Type of "mul(Two(), "a")" is "str"
reveal_type(mul(Two(), 1))  # Type of "mul(Two(), 1)" is "str"  -- huh?
```

The last line is incorrectly inferred, and should be `int` instead, matching the `Two() * 1` outcome. It's as if only the first overload is considered.

Additionally, the following error is reported:

```
Argument of type "Literal[1]" cannot be assigned to parameter "rhs" of type "X@mul" in function "mul"
  "Literal[1]" is not assignable to "str"  (reportArgumentType)
```

---

If the protocol is rewritten, this will behave as expected (for the most part):

```py
class _CanMulWorkaround[F: Callable[[Never], object]](Protocol):
    @property
    def __mul__(self, /) -> F: ...

type CanMulWorkaround[X, Y] = _CanMulWorkaround[Callable[[X], Y]]

def mul_workaround[X, Y](lhs: CanMulWorkaround[X, Y], rhs: X, /) -> Y:
    # return lhs * rhs   <-- This now results in 2 errors (another bug?)
    return lhs.__mul__(rhs)

reveal_type(mul_workaround(Two(), "a"))  # Type of "mul_workaround(Two(), "a")" is "str"
reveal_type(mul_workaround(Two(), 1))  # Type of "mul_workaround(Two(), 1)" is "int"
```

I see no reason why `CanMul` and `CanMulWorkaround` should result in different outcomes. 
And I therefore believe that the correct behavior would be for `CanMul` to *also* take all overloads into consideration, similar to `CanMulWorkaround`.

---

Unfortunately, `CanMulWorkaround` also doesn't work in the following scenario:

```py
class RTwo(Two):
    def __rmul__[X, Y](self: CanMulWorkaround[X, Y], x: X, /) -> Y:
        return self.__mul__(x)

reveal_type("a" * RTwo())  # Type of ""a" * RTwo()" is "str"
reveal_type(int() * RTwo())  # Operator "*" not supported for types "int" and "RTwo"  (reportOperatorIssue)
```

I'm not sure whether this is technically speaking the same issue, but I didn't want to raise more issues than necessary. So let me know if I should open another issue for this.

---

https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgDRQBy1AbtSPQArgxgDGYJemFYgSYIgBN6IFtVIB9eAmoAoFb3IBnTVAAqAdzAAuFVDNQAAsLZjJp8xOrAo8%2BRACuJVwApN1EsD0AB5GUJow7FAA9ACUUAC0AHxhEaEAdBn2ZlYithJZUI7Orh5e8r7%2BgVAhmCgw9LEJyagw6ZnmhU4ubp4%2BfgHBoeEgUAA%2BtfXRcUkpI%2BMtJh0dMjDuIChQAExQAFTVajKsCkrU3gZg3nF7AERE13FQAMR6iNRQYM7X55e7ULfX10wOmuw2uKkOci8JzOhh%2BewAjA9nrpXu9Pt8rlB4YCkMCWmCVI8iWoNERtIQiCgALKeADaAA16ABNAC63i4YB4-BIMUWDi6JV65X6VRqjKmTSgTLaaTURSgpQZzLZJAAFppQsRqXTxazpOrQuLGjNpQUVmsNmqdHsQOqDrJjq9vKUYRcYvR-jEkS9lGi-i6MR67jEccDQeCHVCnQHYe6sV6zMjUR9-Z5XZd6IjQ39QWZ4vEoKr3KqAPyEqAAQRAaHcEGodT9Jz%2BABkkDA2KRafCWYDeJSUJyoBQ3mTNEg0ChqBJYGAoAgiCAiHX2yNrrbNICU03rvSLKUcRtgO4ULwYEgwBtrvvy2Zrq2V53u9mB-hR%2BOUJQaDOcxFAVBvDIOB4FWNZ1nUKLKDEahEo8JJaDo8hajSJAAOq4AA1guYDHhItIAGKaqQ5BUNQtK0kwIgskIFAAFbUKeLJshyXICLyBQWAg4DKHgcAFPKgplBUAwSjMBFQBksoqE2SGeGhICYeAOFKlKLJQAAvC4MmoRhWFKcQZCfqRDJUSpLJyl0pTyIY8m6SguG6iqBoUtq2k2YpdnKXqUDroaDTTMkpodM85rrFAVq-Ou5gADz5noqq4lAA76N51CaJ4MA6KgWxQGw4AgDo3iUpyqpsEO7hoCWUHLNQqyhVaaQCT465QRGRxRsozq9NZCnYXZ6Zxp63oQW8KZXl1OnuRI-VBvc2Ygr%2BrWQoo0bjW5vVTYG8ZDcmnyWd1tkbbGmYhkCfz4mohLEuo8FQAASt85xsR0-HyCAlnyJ5bIioRLlyT1SkOYMUBGv5Up8ksKW1RsIoNT0glBC1EKOh1-y-Pdsbbb6o2o3s6NunN4ZI%2B1pwtHCd0Yt6ADy3FEDwq47ICL5hO4CBAe207ALgsCvHidSApS07XHjf4AdQbPUx2dMAJLaO41BQUAA

---

For some context: I was planning to use this trick to significantly reduce the amount of code duplication in the numpy stubs for the operator methods of `numpy.ndarray` and subclasses of `numpy.generic`. 
And this trick should be flexible enough to define `__divmod__` without any overloads, i.e. `(self: CanMod[X, Y], x: X, /) -> tuple[Y, Y]`.
 examples of where this could make a huge difference:

Some (of the many) use-cases that came to mind:
- `numpy.float64`: https://github.com/numpy/numpy/blob/2eccb5eb23a3b5de302430aa33a41b856b713741/numpy/__init__.pyi#L3899-L3999
- `numpy.ndarray`: https://github.com/numpy/numpy/blob/2eccb5eb23a3b5de302430aa33a41b856b713741/numpy/__init__.pyi#L2555-L3298.
- `scipy.sparse._base._spbase` (base of 14 concrete subtypes `scipy.sparse.(bsr|coo|csc|csr|dia|dok|lil)_(array|matrix)`): https://github.com/jorenham/scipy-stubs/blob/fba8e3fc19e80ca0923ddceec01886dd349f0f6f/scipy-stubs/sparse/_base.pyi#L200-L429

So I'm sure you see that it's not an exaggeration to say that this could save thousands of lines of code in `numpy`, and also in `scipy-stubs`.

**Comment by erictraut** [comment]
In Python, evaluating a call of an overloaded callable is very complex. It's unfortunately not as simple as "matching" a single signature. This complexity brings with it many limitations and undesirable side effects, but we're unfortunately stuck with this complexity because mypy established an (not-well-documented) set of behaviors that everyone now depends upon. See [here](https://discuss.python.org/t/draft-typing-spec-chapter-for-overloads/60788/10) for my attempt to document and standardize this behavior. 

This complexity means that it's not feasible to do what you're proposing here. Constraint solving is already complex, and the constraint-solving behaviors are not defined in the spec. What you're proposing here would involve standardizing behaviors between overload resolution and constraint solving, and I don't think that's going to happen. If you'd like to attempt it, you're welcome to try. Without such standardization, you wouldn't be able to rely on behaviors across type checkers.

I'll also point out that evaluating the type of binary operators requires significant logic within type checkers to mirror the Python runtime behavior of these operators. For example, the `*` operator invokes code at runtime that is significantly more complex than simply calling through to the `__mul__` method on the first operand.

**Comment by jorenham** [comment]
Are you saying that the following result is the intended behavior, @erictraut?

```py
from typing import Literal, Protocol, overload

class A:
    @overload
    def f(self, x: Literal[0], /) -> Literal[1]:...
    @overload
    def f(self, x: int, /) -> int:...
    def f(self, x: int, /) -> int:
        return x or 1

class CanF[X, R](Protocol):
    def f(self, x: X, /) -> R: ...

def f[R, X](s: CanF[X, R], x: X, /) -> R:
    return s.f(x)

a: A = A()

reveal_type(a.f(0))  # Literal[1]
reveal_type(f(a, 0))  # int

reveal_type(a.f(1))  # int
reveal_type(f(a, 1))  # Literal[1] + Error (reportArgumentType)
```

**Comment by erictraut** [comment]
> Are you saying that the following result is the intended, or acceptable, behavior, @erictraut?

Yes, that's what I'm saying. You'll see that mypy also emits a similar error for this code.

When evaluating the call `f(a, 1)`, a type checker evaluates the value of the first argument, which is `A`. It then must solve values for `R` and `X` for the corresponding parameter whose type is `CanF[X, R]`. It does this by protocol matching. When it comes to the `f` method, the solution is ambiguous, so type checkers like mypy and pyright use heuristics. In this case, the heuristic is to use the first overload whose arity matches the protocol's (non-overloaded) method. One might think that a type checker would try all possible overloads, but this would be impractical, especially if a protocol has multiple methods. There would be a combinatoric explosion of potential solutions to try.

**Comment by jorenham** [comment]
By redefining `CanF` as 

```py
class HasF[F: Callable[[Never], object]](Protocol):
    @property
    def f(self) -> F: ...

type CanF[X, R] =  HasF[Callable[[X], R]]
```

we get a different, result:

```
reveal_type(a.f(0))  # Literal[1]
reveal_type(f(a, 0))  # int

reveal_type(a.f(1))  # int
reveal_type(f(a, 1))  # int, no error
```

https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgDRQAySM1Ip9ActQG6v0AK4GGADGYEvTC8QJMEQAmAKEUjyAZzVQAggC5FUA1AACU1rIX7D86sCjAAFGuolg9AB47GzVqQDaABgBdegB6AEooAFoAPi8WNhJfAEZAnQA6DMsDE2lzJUMoa1sHJxd3T1QYUIiYzBQYdMyCortHZ1coDzqqqHCo2Mq9AoKQahgAVxAUTqhcKCTlVSINKAAJZYAxXw3PYjJKGl9fLmlg2YoAK2oRGEDA%2B0EwYTESMKHDIwRwBFZ4LMKbK1SsAarEdlAMmllPAfoQiCgtgANegAJUCUAAvAZ1motntyFRqEdEWc0YFlC1gL4UfQSY5dvCkaizl1kb1QVAUe8DKMJlMoGo0g43GFlERPFpMdp7KLFKNeKQAPow6j2IhC%2Bz%2BMIRKAAYjiPkSKTlPGoSpV9gcRHoWp1%2BsqynlZpIysQqvVDiS2oM9vqJoVLotVvoXrt3XoKDAUFY4BAiiAA

This is a more *useful* result, and closer to what I expected it to be. I can't think of any situations where the previous outcome is more helpful that this outcome. 

So can the previous `CanF` protocol be made to behave like in this case @erictraut?

**Comment by erictraut** [comment]
> So can the previous CanF protocol be made to behave like in this case @erictraut?

Are you asking whether pyright's behavior could be modified such that it behaves the way you want in this case? Or are you asking whether the typing spec could be amended to specify the behavior such that you could rely on this behavior within a public type stub?

As I said above, the rules for constraint solving are not currently specified in the typing spec. Type checkers use various internal heuristics to solve type variables in call expressions. Even if I were to change pyright's behavior here, mypy and other type checkers would behave differently. 

If you would like to see this behavior standardized across type checkers, you're welcome to push for standardization, but it will be a *very* heavy lift and likely take years before you reach resolution and get all type checkers to adjust their implementations accordingly. 

I'm not inclined to modify pyright's heuristics here unless/until there is a specification. Doing so will create churn for pyright users and will not help stub authors in general because they still won't be able to rely on the (unspecified) behavior.

**Comment by jorenham** [comment]
> Are you asking whether pyright's behavior could be modified such that it behaves the way you want in this case? 

Yes

> Or are you asking whether the typing spec could be amended to specify the behavior such that you could rely on this behavior within a public type stub?

I would like that, but as far as I can tell, this difference in `Protocol` behavior isn't part of the typing spec.

> Even if I were to change pyright's behavior here, mypy and other type checkers would behave differently.

Yes, and I plan to address this with mypy as well. And I think my chances of success will be significantly higher if pyright were to take the lead in this. And I'm sure that you can understand why Pyright is the first door I knock at.

> If you would like to see this behavior standardized across type checkers, you're welcome to push for standardization, but it will be a _very_ heavy lift and likely take years before you reach resolution and get all type checkers to adjust their implementations accordingly.

I'm hoping that the quickest road to standardization of currently unspecified behaviour, is by setting a good example. In my estimation everyone in this area looks up to pyright (and for good reason). So by starting here, the path towards standardization will be much smoother and shorter.
However, I'm afraid that converse may also be true.

> I'm not inclined to modify pyright's heuristics here unless/until there is a specification. Doing so will create churn for pyright users and will not help stub authors in general because they still won't be able to rely on the (unspecified) behavior.

As a stub author myself, I can guarantee you that this will help *a lot*. Like I said before, I plan to raise this with mypy as well. And if you could take the lead here, then I expect my chances of success to be significantly higher, and efficiently achieve the same goal.

**Comment by jorenham** [comment]
> Right now, pyright and mypy (apparently) choose the first overload when multiple overloads match during protocol matching of a generic protocol. Are you suggesting that type checkers should choose the _last_ overload instead? Or something else?

I want this `CanF` 

```pyi
class CanF[X, R](Protocol):
    def f(self, x: X, /) -> R: ...
```

to consider all of `f`'s overloads, in the same way that this other `CanF` (from https://github.com/microsoft/pyright/issues/9663#issuecomment-2578096442) does:

```pyi
class HasF[F: Callable[[Never], object]](Protocol):
    @property
    def f(self) -> F: ...

type CanF[X, R] =  HasF[Callable[[X], R]]
```

So this last `CanF` already behaves like I want the first `CanF` behaves; that's all.

**Comment by erictraut** [comment]
I don't know what you mean by "consider all of f's overloads". How would you spec that? 

Protocol matching works by assigning attributes and methods that appear within the protocol to determine whether the target class is compatible, so this comes down to assignment and constraint solving behaviors.

In the case of `CanF` in your example, `A.f` has two overloads that can be assigned to `CanF.f`. The assignability rules for callables say that an overloaded callable is assignable to another (non-overloaded) callable if one or more of its overload signatures is assignable. In this case, this assignability criterion is met, but we need to decide which of the two overload signature to use for purposes of constraint solving. We can either choose the first or the second (last) because both of them are assignable. The constraint solver needs a set of concrete constraints to solve.

**Comment by jorenham** [comment]
I have streamlined the example:

```py
from typing import Literal as L, Protocol, final, overload


### Minimal structural typing

class CanF[X, R](Protocol):
    def f(self, x: X, /) -> R: ...

def call_f[X, R](obj: CanF[X, R], arg: X) -> R:
    return obj.f(arg)


### Implementation

@final
class A:
    @overload
    def f(self, x: L[0], /) -> str:...
    @overload
    def f(self, x: L[1], /) -> int:...
    def f(self, x: L[0, 1], /) -> int | str:
        return x or str(x)

a = A()


### Validate the protocol

b0: CanF[L[0], str] = a  # accepted
b1: CanF[L[1], int] = a  # accepted


### Demo of the expected outcome

reveal_type(a.f(0))  # str
reveal_type(a.f(1))  # int


### Actual outcome

reveal_type(call_f(a, 0))  # str
# Argument of type "A" cannot be assigned to parameter "obj" of type "CanF[X@call_f, R@call_f]" in function "call_f"
#  "A" is incompatible with protocol "CanF[X@call_f, R@call_f]"
#    "f" is an incompatible type
#      No overloaded function matches type "(x: X@CanF, /) -> R@CanF"  (reportArgumentType)
reveal_type(call_f(a, 1))  # str
# Argument of type "Literal[1]" cannot be assigned to parameter "arg" of type "X@call_f" in function "call_f"
#  "Literal[1]" is not assignable to type "Literal[0]"  (reportArgumentType)
```

[pyright-play](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2U5AzoQDRQAK4MYAxmJU8KlUzAA3MpTDkAJgCgZAYnlQAsqixUotGCACuHGForV4ydDI6U69AMLkUAMQDaADSYAlALoAKVmHZdKASgAuKShQqAkSYChgD1oSSmAmAA9AqGcoAHp-KABaAD4oF1SAOlKZCKiOKkoAfWAnV08wACMAK1TrOwbCtyZyEDRUx2z8wuCwqBASPRAUKBbW4pj%2BtH8ZKXlZKABJbEoSCBIUGHIYJDAUGQABPhQqKTMLKABBcbCr4VFxaQmK6Nj4okoClCPYAAy9TIjAoaECBUrFELvT4gMSSJGhP4xOIJZKpAj2ACMkKyuQKqBg8LKv0i-xxQJBBLBTGJTFJowpUAAPupNG8JqEpjM5kl5iBeSAPEk1lJyFAALwvDwyjYKABqVCQElOJFgAAtdQg2JxuDJmmCOjYHEzIbC3AqaKEtuQOBwSAhSNJmoTLV0CazMMd7Yq5VBna73Z71psoAARA5geZRGAGqAkJIIEi6EgSeZaGBcQ4yKYiKg1IwkDzkJYeMH%2BbJhiVSEskMsVqs1wn1p2BmDRhTPXRaNRgfOFkjFkil2rtqqUWrLJh1htbWEbF4DLSHY5J2CIXUAImeB6gVRQKB8UGauosSDQKBzsETCH65EOpHFB4WJ7Ayf3UAPToHEcK45wXVxQOqOo3BPVBoi0FBdHOOYDzAuoD3XADj0wehUELF8zmafYoAAd2IPUoCNHwTWoQCrScSD5zqCC0OAGDMNCA9gFg%2BgbEDfDTiQIjdQrDiwgAOUTFE0QiXNgAQpCLigCBTg4A16ArACpSGK4gLZaFCl0q0TygDwphwPBnk3bcYAAFX3NYWzbfcPFYqsWW7Rs1y2Ky0C3I58F-PdMwAogPyoIkYNPGwL3wa8aFoWg7wfXN2Eo193zIACVh-P8QoPEDWNguZ5MQs4lNQqDuMwg8wrICLiR4qBYoSpL73IYSn2Cw86oMcEotM8zcBgXz-OOezMzWIA)


1. The `call_f(a, 0)` error reports that `a: A` cannot be assigned to `CanF`. This is incorrect, because the `b0: CanF[L[0], str] = a` is a valid assignment.
2. The `call_f(a, 1)` error reports that `L[1]` cannot be assigned to `arg: X`, where `X@call_f` is bound to `X@x` in `A.f`.
So given `CanF[X, R] ~ A`, we either have `X: L[0] -> R: str` or `X: L[1] -> R: int` in `(CanF[X, R] ~ A).f`.
Therefore, `arg: L[1]` in `call_f` implies `obj: CanF[L[1], R]`, and `CanF[X, R] ~ A` implies `X: L[1] -> R: int`, which gives us `obj: CanF[L[1], int]`.
The return type `R@can_f` is therefore that very same `int`. Additionally, `arg: L[1]` should be accepted, so the reported error is a false positive.

These seem like valid & solvable constraints to me. 
I'm not sure whether the "type-mapping" form that I expressed them as, fits within pyright's constraint-solver framework, as it requires "compressing" multiple overloads into a signature of causally correlated parameter types.
But I can imagine that it's significantly easier to deal with a single signature and correlated parameter types, than with multiple correlated signatures of uncorrelated parameter types.

---


> How would you spec that?

I'm not sure what level of formality you were thinking of, so I hope that by now I've given you enough of an idea on what I think could and should work here :)

**Comment by jorenham** [comment]
> If we did, I think it's safe to say that what you're trying to do here would be disallowed.

why?
